contrib/mesos/pkg/minion/tasks/task.go

```diff
@@ -241,61 +251,75 @@ type exitError interface {
-		sendOnce.Do(func() {
-			t.tryComplete(wr)
-		})
-	}
 	// listen for normal process completion in a goroutine; don't block because we need to listen for shouldQuit
 	waitCh := make(chan *Completion, 1)
 	go func() {
-		wr := &Completion{}
+		wr := &Completion{name: t.name}
 		defer func() {
 			waitCh <- wr
 			close(waitCh)
 		}()
 
-		if err := waiter(); err != nil {
+		if err := t.cmd.Wait(); err != nil {
 			if exitError, ok := err.(exitError); ok {
 				if waitStatus, ok := exitError.Sys().(syscall.WaitStatus); ok {
-					wr.name = t.name
 					wr.code = waitStatus.ExitStatus()
 					return
 				}
 			}
 			wr.err = fmt.Errorf("task wait ended strangely for %q: %v", t.bin, err)
-		} else {
-			wr.name = t.name
 		}
 	}()
 
-	// wait for the process to complete, or else for a "quit" signal on the task (at which point we'll attempt to kill manually)
-	select {
-	case <-t.shouldQuit:
-		// check for tie
+	t.tryComplete(t.awaitDeath(defaultKillGracePeriod, waitCh))
+	return taskShouldRestart
+}
+
+// awaitDeath waits for the process to complete, or else for a "quit" signal on the task-
+// at which point we'll attempt to kill manually.
+func (t *Task) awaitDeath(gracePeriod time.Duration, waitCh <-chan *Completion) (wr *Completion) {
+	wr = &Completion{name: t.name, err: fmt.Errorf("failed to kill process: %q", t.bin)}
+	forceKill := false
+
+	// the loop is here in case we receive a shouldQuit signal; we need to kill the task.
+	// in this case, first send a SIGTERM (force=false) to the task and then wait for it
+	// to die (within the gracePeriod). if it doesn't die, then we loop around only this
+	// time we'll send a SIGKILL (force=true) and wait for a reduced gracePeriod. There
+	// does exist a slim chance that the underlying wait4() syscall won't complete before
+	// this process dies, in which case a zombie will rise. Starting the mesos slave with
+	// pid namespace isolation should mitigate this.
+waitLoop:
+	for i := 0; i < 2; i++ {
 		select {
-		case wr := <-waitCh:
-			// we got a signal to quit, but we're already finished; attempt best effort delvery
-			trySend(wr)
-		default:
-			// Wait() has not exited yet, kill the process
-			log.Infof("killing %s : %s", t.name, t.bin)
-			pid, err := t.cmd.Kill()
+		case <-t.shouldQuit:
+			select {
+			case wr = <-waitCh:
+				// got a signal to quit, but we're already finished
+				break waitLoop
+			default:
+			}
+
+			log.Infof("killing %s (force=%t) : %s", t.name, forceKill, t.bin)
+			pid, err := t.killFunc(forceKill)
 			if err != nil {
-				trySend(&Completion{err: fmt.Errorf("failed to kill process: %q pid %d: %v", t.bin, pid, err)})
+				log.Warningf("failed to kill process: %q pid %d: %v", t.bin, pid, err)
+				break waitLoop
+			}
+
+			// Wait for the kill to be processed, and child proc resources cleaned up; try to avoid zombies!
+			select {
+			case wr = <-waitCh:
+				break waitLoop
+			case <-time.After(gracePeriod):
+				gracePeriod /= 2
```

Comment:
I want a timeout, but a shorter one than we used initially. / 2 is deterministic and yields the desirable effect.